torch/autograd/grad_mode.py

```diff
@@ -1,8 +1,36 @@
 import torch
 import functools
+import inspect
 
+class _DecoratorContextManager:
+    """Allow a context manager to be used as a decorator"""
 
-class no_grad(object):
+    def __call__(self, func):
+        if inspect.isgeneratorfunction(func):
+            return self._wrap_generator(func)
+
+        @functools.wraps(func)
+        def decorate_context(*args, **kwargs):
+            with self:
+                return func(*args, **kwargs)
+        return decorate_context
+
+    def _wrap_generator(self, func):
+        """Wrap each generator invocation with the context manager"""
+        @functools.wraps(func)
+        def generator_context(*args, **kwargs):
+            gen = func(*args, **kwargs)
+            while True:
+                try:
+                    with self:
+                        x = next(gen)
```

Comment:
Remove the temporary `x`:

```python
with self:
    yield next(gen)
```

Otherwise you keep an extra value alive unnecessarily, which in PyTorch could be a large Tensor.

Reply:
Never mind — that’s not going to work

Reply:
Yes, I wrote that at first but the `yield` statement must be outside of the context manager to not leak into the calling context.